html/semantics/forms/the-input-element/type-change-state.html is failing in WebKit

cdumez opened the pull request:
#### 3763b9dfd24051056b3218d862d8a908b8f5ab0e
<pre>
html/semantics/forms/the-input-element/type-change-state.html is failing in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=243311">https://bugs.webkit.org/show_bug.cgi?id=243311</a>

Reviewed by Darin Adler.

Per the specification [1]:
```
If the previous state of the element&apos;s type attribute put the value IDL
attribute in the value mode, and the element&apos;s value is not the empty string,
and the new state of the element&apos;s type attribute puts the value IDL attribute
in either the default mode or the default/on mode, then set the element&apos;s value
content attribute to the element&apos;s value.
```

In our implementation, we were missing the `and the element&apos;s value is not the
empty string` part of the check. As a result, when changing the type from
&quot;number&quot; to &quot;checkbox&quot; (for example) and if m_valueIfDirty was the emptyString,
we would set the &quot;value&quot; content attribute to the empty string. In general, this
is not very observable because input.value would return the empty string if
whether the &quot;value&quot; content attribute is missing or the empty string. However,
for input types such as &quot;checkbox&quot; and &quot;radio&quot;, it is observable since
`input.value` is expected to &quot;on&quot; when the &quot;value&quot; content attribute is missing.
This bug would cause `input.value` to return &quot;&quot; instead of &quot;on&quot; in such case.

[1] <a href="https://html.spec.whatwg.org/#the-input-element">https://html.spec.whatwg.org/#the-input-element</a>:attr-input-type-37 (step 1)

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/type-change-state-expected.txt:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):

Canonical link: <a href="https://commits.webkit.org/252967@main">https://commits.webkit.org/252967@main</a>
</pre>

darinadler commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -525,8 +525,10 @@ void HTMLInputElement::updateType()
+    if (didStoreValue && !willStoreValue && hasDirtyValue()) {
+        if (auto dirtyValue = std::exchange(m_valueIfDirty, { }); !dirtyValue.isEmpty())
+            setAttributeWithoutSynchronization(valueAttr, AtomString { WTFMove(dirtyValue) });
+    }
```
Since we are checking `!isEmpty()`, we don’t need the `hasDirtyValue()` check; can just take that out. Also wondering if any other call sites making `hasDirtyValue()` checks also want to handle empty dirty values differently.

cdumez committed:
html/semantics/forms/the-input-element/type-change-state.html is failing in WebKit
https://bugs.webkit.org/show_bug.cgi?id=243311

Reviewed by Darin Adler.

Per the specification [1]:
```
If the previous state of the element's type attribute put the value IDL
attribute in the value mode, and the element's value is not the empty string,
and the new state of the element's type attribute puts the value IDL attribute
in either the default mode or the default/on mode, then set the element's value
content attribute to the element's value.
```

In our implementation, we were missing the `and the element's value is not the
empty string` part of the check. As a result, when changing the type from
"number" to "checkbox" (for example) and if m_valueIfDirty was the emptyString,
we would set the "value" content attribute to the empty string. In general, this
is not very observable because input.value would return the empty string if
whether the "value" content attribute is missing or the empty string. However,
for input types such as "checkbox" and "radio", it is observable since
`input.value` is expected to "on" when the "value" content attribute is missing.
This bug would cause `input.value` to return "" instead of "on" in such case.

[1] https://html.spec.whatwg.org/#the-input-element:attr-input-type-37 (step 1)

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/type-change-state-expected.txt:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):

Canonical link: https://commits.webkit.org/252967@main
